### PR TITLE
Upsert

### PR DESCRIPTION
**Pull request by MDunitz**
## Need
- gracefully handle insertion of the same data

## Approach
- add rule to the table in the db
- if the row already exists do nothing

## Testing
- manual testing in db
- unit tests pass even with removal of _db_ignore_insert_conflicts event listener

## Todo
- reinit db
- reload data

### CODE REVIEW

File: dcpquery/db/__init__.py [hunk 1]
```diff
@@ -174,8 +136,126 @@ def create_recursive_process_functions_in_db():
...
+class ProcessProcessLink(SQLAlchemyBase):
+    __tablename__ = 'process_join_table'
+    id = Column(Integer, primary_key=True)
+    child_process_uuid = Column(UUID, ForeignKey("processes.process_uuid"))
+    parent_process_uuid = Column(UUID, ForeignKey("processes.process_uuid"))
+    parent_process = relationship(Process, foreign_keys=[parent_process_uuid])
+    child_process = relationship(Process, foreign_keys=[child_process_uuid])
+
+
+class InitDB:
```
Comment (kislyuk) [inline dcpquery/db/__init__.py]: I don't think this pattern works well semantically. The class should normally represent a type of thing, not an action. If you want to use a class to organize this, consider the following pattern instead:
```
class DCPQueryDBManager:
    def init_db(self, dry_run=True):
        ...

DCPQueryDBManager().init_db()
```

File: dcpquery/db/__init__.py [hunk 2]
```diff
@@ -174,8 +136,126 @@ def create_recursive_process_functions_in_db():
...
+                            ON INSERT TO process_file_join_table
+                                WHERE EXISTS (
+                                  SELECT 1
+                                FROM process_file_join_table
+                                WHERE process_uuid = NEW.process_uuid
+                                AND process_file_connection_type=NEW.process_file_connection_type
+                                AND file_uuid=NEW.file_uuid
+                            )
+                            DO INSTEAD NOTHING;
+                        """
```
Comment (kislyuk) [inline dcpquery/db/__init__.py]: Please format the closing quotes to be at a consistent indent. My preference is for them to be at the same indent as the variable assignment line (line 167 here, 4 spaces).

Similarly, please format the SQL to be at a fixed indent. My preference is to indent it one level in from the variable assignment line (so here, line 180 would be indented 8 spaces).

File: dcpquery/db/__init__.py [hunk 3]
```diff
@@ -174,8 +136,126 @@ def create_recursive_process_functions_in_db():
...
+                                ON process_join_table.child_process_uuid = recursive_table.parent_process_uuid)
+                            SELECT * from recursive_table;
+                            $$ LANGUAGE SQL;
+"""
+
+    def create_upsert_rules_in_db(cls):
+        config.db_session.execute(
+            cls.upsert_bundle_file_link_rule_sql + cls.upsert_bundle_rule_sql + cls.upsert_file_rule_sql
+        )
+        # TODO do I need to create an upsert rule for bundle_file_links or process_join_table?
```
Comment (kislyuk) [inline dcpquery/db/__init__.py]: Can this TODO be resolved? If not, can you rephrase it like this:
```
# TODO (MDunitz) need to create an upsert rule for bundle_file_links or process_join_table?
```

File: dcpquery/db/__init__.py (hunk 1 above)
Comment (MDunitz) [reply dcpquery/db/__init__.py]: 👍

File: dcpquery/db/__init__.py (hunk 2 above)
Comment (MDunitz) [reply dcpquery/db/__init__.py]: 👍

File: dcpquery/db/__init__.py (hunk 3 above)
Comment (MDunitz) [reply dcpquery/db/__init__.py]: 👍
